include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h

```diff
@@ -89,56 +99,55 @@ struct sycl_iterator
     {
         return it - forward;
     }
+    template <access_mode inMode>
     difference_type
-    operator-(const sycl_iterator& it) const
+    operator-(const sycl_iterator<inMode, T, Allocator>& it) const
     {
-        assert(buffer == it.get_buffer());
-        return idx - it.idx;
+        return eval_diff_with(it);
     }
+    template <access_mode inMode>
     bool
-    operator==(const sycl_iterator& it) const
+    operator==(const sycl_iterator<inMode, T, Allocator>& it) const
     {
-        assert(buffer == it.get_buffer());
-        return *this - it == 0;
+        return eval_diff_with(it) == 0;
     }
+    template <access_mode inMode>
     bool
-    operator!=(const sycl_iterator& it) const
+    operator!=(const sycl_iterator<inMode, T, Allocator>& it) const
     {
-        assert(buffer == it.get_buffer());
-        return !(*this == it);
+        return eval_diff_with(it) != 0;
     }
+    template <access_mode inMode>
     bool
-    operator<(const sycl_iterator& it) const
+    operator<(const sycl_iterator<inMode, T, Allocator>& it) const
     {
-        assert(buffer == it.get_buffer());
-        return *this - it < 0;
+        return eval_diff_with(it) < 0;
     }
 
     sycl::buffer<T, dim, Allocator>
     get_buffer() const
     {
         return buffer;
     }
-};
 
-// mode converter when property::noinit present
-template <access_mode __mode>
-struct _ModeConverter
-{
-    static constexpr access_mode __value = __mode;
+    Size
+    get_index() const
```

Comment:
Why do we introduce new public method?

Reply:
It's required for cases when we make some operation with sycl_iterator with different access mode: for example, call ```sycl_iterator<access_mode::read_write>().operator==(sycl_const_iterator());```